File: frontend/src/types/escort.ts
export interface EscortCard {
  id: string
  stage_name: string
  slug: string
  age: number | null
  nationality: string | null
  ethnicity: string | null
  borough_name: string | null
  borough_slug: string | null
  availability_type: string | null
  rate_1hour: number | null
  subscription_tier: 'free' | 'essential' | 'premium' | 'elite'
  verification_level: 0 | 1 | 2 | 3
  blue_tick_active: boolean
  available_now: boolean
  std_tested: boolean
  primary_photo_url: string | null
  service_tags: string[]
  profile_type: 'individual' | 'couple'
}

export interface Photo {
  id: string
  url: string
  thumbnail_url: string | null
  is_primary: boolean
  sort_order: number
}

export interface EscortProfile {
  id: string
  stage_name: string
  slug: string
  age: number | null
  nationality: string | null
  ethnicity: string | null
  height_cm: number | null
  build: string | null
  hair_colour: string | null
  eye_colour: string | null
  dress_size: string | null
  chest: string | null
  borough_name: string | null
  borough_slug: string | null
  availability_type: string | null
  rate_30min: number | null
  rate_1hour: number | null
  rate_2hours: number | null
  rate_overnight: number | null
  about_me: string | null
  languages: string[] | null
  booking_notice: string | null
  std_tested: boolean
  std_tested_date: string | null
  subscription_tier: string
  verification_level: number
  blue_tick_active: boolean
  available_now: boolean
  profile_views: number
  service_tags: string[]
  photos: Photo[]
  created_at: string
  profile_type: 'individual' | 'couple'
  whatsapp_number: string | null
  phone_number: string | null
}

export interface EscortDashboard extends EscortProfile {
  email: string
  borough_id: string | null
  rate_30min: number | null
  rate_2hours: number | null
  rate_overnight: number | null
  is_email_verified: boolean
  profile_complete: boolean
  subscription_expires_at: string | null
  contact_clicks: number
  photo_limit: number
  whatsapp_number: string | null
  phone_number: string | null
  blue_tick_active: boolean
  blue_tick_stripe_subscription_id: string | null
  stripe_subscription_id: string | null
  profile_type: 'individual' | 'couple'
}

export interface Borough {
  id: string
  name: string
  slug: string
  description: string | null
  seo_title: string | null
  seo_description: string | null
  is_premium_area: boolean
  escort_count: number
}

export interface PaginatedResponse<T> {
  items: T[]
  total: number
  page: number
  per_page: number
  pages: number
}

export interface SearchFilters {
  borough_slug?: string
  ethnicity?: string
  availability_type?: string
  profile_type?: string
  min_age?: number
  max_age?: number
  min_rate?: number
  max_rate?: number
  std_tested?: boolean
  available_now?: boolean
  blue_tick_only?: boolean
  service_tag?: string
  page?: number
  per_page?: number
}

export const SERVICE_TAGS = [
  'GFE', 'PSE', 'OWO', 'OWO-CIM', 'CIM', 'CIMWS', 'COB', 'CIF',
  'Massage', 'Erotic Massage', 'Tantric Massage', 'Body Slide', 'Nude Massage',
  'Happy Ending', 'Hand Relief', 'DFK', '69', 'Anal', 'A-Level',
  'Deep Throat', 'BBBJ', 'Rimming', 'Squirting',
  'BDSM', 'Bondage', 'Domination', 'Submission', 'Fetish', 'Role Play',
  'Tie & Tease', 'S&M', 'Watersports', 'CBT', 'Strapon',
  'Duo', 'Threesome', 'Couples Welcome',
  'Dinner Date', 'Travel Companion', 'Overnight',
  'Striptease', 'Lap Dance', 'Foot Fetish', 'Shower Together',
  'Toys', 'Party Friendly', 'Webcam',
] as const

export const ETHNICITIES = [
  'White', 'Black', 'Asian', 'Latin', 'Mixed', 'Middle Eastern', 'Indian', 'Other',
] as const

export const BUILD_TYPES = [
  { value: 'slim', label: 'Slim' },
  { value: 'athletic', label: 'Athletic' },
  { value: 'curvy', label: 'Curvy' },
  { value: 'petite', label: 'Petite' },
  { value: 'bbw', label: 'BBW' },
] as const

export const LANGUAGES = [
  'English', 'Romanian', 'Spanish', 'Russian', 'Polish', 'Portuguese',
  'Brazilian Portuguese', 'Thai', 'Hungarian', 'Italian', 'French', 'German',
  'Arabic', 'Mandarin', 'Japanese', 'Korean', 'Turkish', 'Greek',
  'Swedish', 'Dutch', 'Czech', 'Ukrainian', 'Albanian', 'Bulgarian',
  'Serbian', 'Lithuanian', 'Hindi', 'Punjabi', 'Urdu', 'Persian',
  'Bengali', 'Somali', 'Other',
] as const

export const SUBSCRIPTION_TIERS = {
  free: { label: 'Free', price: 0, color: 'stone' },
  essential: { label: 'Essential', price: 2499, color: 'silver' },
  premium: { label: 'Premium', price: 4999, color: 'gold' },
  elite: { label: 'Elite', price: 8999, color: 'gold' },
} as const

File: backend/app/schemas/escort.py
import re
import uuid
from datetime import datetime
from pydantic import BaseModel, field_validator
from typing import Optional

_PHONE_RE = re.compile(r'^\+?[\d\s\-().]{7,25}$')


class PhotoOut(BaseModel):
    id: uuid.UUID
    url: str
    thumbnail_url: Optional[str]
    is_primary: bool
    sort_order: int

    model_config = {"from_attributes": True}


class EscortCardOut(BaseModel):
    """Lightweight schema for grid/listing views."""
    id: uuid.UUID
    stage_name: str
    slug: str
    age: Optional[int]
    nationality: Optional[str]
    ethnicity: Optional[str]
    borough_name: Optional[str] = None
    borough_slug: Optional[str] = None
    availability_type: Optional[str]
    rate_1hour: Optional[int]
    subscription_tier: str
    verification_level: int
    blue_tick_active: bool = False
    is_founding_member: bool = False
    available_now: bool
    std_tested: bool
    primary_photo_url: Optional[str]
    service_tags: list[str] = []
    profile_type: str = "individual"

    model_config = {"from_attributes": True}


class EscortProfileOut(BaseModel):
    """Full profile schema for individual profile pages."""
    id: uuid.UUID
    stage_name: str
    slug: str
    age: Optional[int]
    nationality: Optional[str]
    ethnicity: Optional[str]
    height_cm: Optional[int]
    build: Optional[str]
    hair_colour: Optional[str]
    eye_colour: Optional[str]
    dress_size: Optional[str]
    chest: Optional[str]
    borough_name: Optional[str] = None
    borough_slug: Optional[str] = None
    availability_type: Optional[str]
    rate_30min: Optional[int]
    rate_1hour: Optional[int]
    rate_2hours: Optional[int]
    rate_overnight: Optional[int]
    about_me: Optional[str]
    languages: Optional[list[str]]
    booking_notice: Optional[str]
    std_tested: bool
    std_tested_date: Optional[str]
    subscription_tier: str
    verification_level: int
    blue_tick_active: bool = False
    is_founding_member: bool = False
    available_now: bool
    profile_views: int
    service_tags: list[str] = []
    photos: list[PhotoOut] = []
    created_at: datetime
    profile_type: str = "individual"
    whatsapp_number: Optional[str] = None
    phone_number: Optional[str] = None

    model_config = {"from_attributes": True}


class EscortUpdateRequest(BaseModel):
    stage_name: Optional[str] = None
    age: Optional[int] = None
    nationality: Optional[str] = None
    ethnicity: Optional[str] = None
    height_cm: Optional[int] = None
    build: Optional[str] = None
    hair_colour: Optional[str] = None
    eye_colour: Optional[str] = None
    dress_size: Optional[str] = None
    chest: Optional[str] = None
    borough_id: Optional[uuid.UUID] = None
    availability_type: Optional[str] = None
    rate_30min: Optional[int] = None
    rate_1hour: Optional[int] = None
    rate_2hours: Optional[int] = None
    rate_overnight: Optional[int] = None
    about_me: Optional[str] = None
    languages: Optional[list[str]] = None
    booking_notice: Optional[str] = None
    std_tested: Optional[bool] = None
    std_tested_date: Optional[str] = None
    service_tags: Optional[list[str]] = None
    available_now: Optional[bool] = None
    profile_type: Optional[str] = None
    whatsapp_number: Optional[str] = None
    phone_number: Optional[str] = None

    @field_validator("whatsapp_number", "phone_number")
    @classmethod
    def validate_phone(cls, v: Optional[str]) -> Optional[str]:
        if v and v.strip():
            v = v.strip()
            if not _PHONE_RE.match(v):
                raise ValueError("Phone number must be in international format, e.g. +44 7700 900000")
            if len(v) > 25:
                raise ValueError("Phone number too long")
        return v or None

    @field_validator("about_me")
    @classmethod
    def cap_about_me(cls, v: Optional[str]) -> Optional[str]:
        if v and len(v) > 600:
            return v[:600]
        return v

    @field_validator("age")
    @classmethod
    def validate_age(cls, v: Optional[int]) -> Optional[int]:
        if v is not None and (v < 18 or v > 99):
            raise ValueError("Age must be between 18 and 99")
        return v


class EscortDashboardOut(BaseModel):
    """Full escort object returned to authenticated escort for dashboard."""
    id: uuid.UUID
    email: str
    stage_name: str
    slug: str
    age: Optional[int]
    nationality: Optional[str]
    ethnicity: Optional[str]
    height_cm: Optional[int]
    build: Optional[str]
    hair_colour: Optional[str]
    eye_colour: Optional[str]
    dress_size: Optional[str]
    chest: Optional[str]
    borough_id: Optional[uuid.UUID]
    borough_name: Optional[str] = None
    availability_type: Optional[str]
    rate_30min: Optional[int]
    rate_1hour: Optional[int]
    rate_2hours: Optional[int]
    rate_overnight: Optional[int]
    about_me: Optional[str]
    languages: Optional[list[str]]
    booking_notice: Optional[str]
    std_tested: bool
    std_tested_date: Optional[str]
    is_email_verified: bool
    verification_level: int
    profile_complete: bool
    available_now: bool
    subscription_tier: str
    subscription_expires_at: Optional[datetime]
    blue_tick_active: bool = False
    is_founding_member: bool = False
    blue_tick_stripe_subscription_id: Optional[str] = None
    stripe_subscription_id: Optional[str] = None
    profile_views: int
    contact_clicks: int
    service_tags: list[str] = []
    photos: list[PhotoOut] = []
    photo_limit: int
    created_at: datetime
    profile_type: str = "individual"
    whatsapp_number: Optional[str] = None
    phone_number: Optional[str] = None
    referral_code: Optional[str] = None

    model_config = {"from_attributes": True}


class EscortSearchParams(BaseModel):
    borough_slug: Optional[str] = None
    ethnicity: Optional[str] = None
    availability_type: Optional[str] = None
    profile_type: Optional[str] = None
    min_age: Optional[int] = None
    max_age: Optional[int] = None
    min_rate: Optional[int] = None
    max_rate: Optional[int] = None
    std_tested: Optional[bool] = None
    available_now: Optional[bool] = None
    blue_tick_only: Optional[bool] = None
    service_tag: Optional[str] = None
    page: int = 1
    per_page: int = 24
